templates/node/src/client.ts.twig

```diff
@@ -0,0 +1,258 @@
+        super(message);
+        this.name = '{{spec.title | caseUcfirst}}Exception';
+        this.message = message;
+        this.code = code;
+        this.type = type;
+        this.response = response;
+    }
+}
+
+class Client {
+    static CHUNK_SIZE = 1024 * 1024 * 5;
+
+    config = {
+        endpoint: '{{ spec.endpoint }}',
+        endpointRealtime: '',
+        selfSigned: false,
+        {%~ for header in spec.global.headers %}
+        {{ header.key | caseLower }}: '',
+        {%~ endfor %}
+    };
+    headers: Headers = {
+        'x-sdk-name': '{{ sdk.name }}',
+        'x-sdk-platform': '{{ sdk.platform }}',
+        'x-sdk-language': '{{ language.name | caseLower }}',
+        'x-sdk-version': '{{ sdk.version }}',
+        {%~ for key,header in spec.global.defaultHeaders %}
+        '{{key}}': '{{header}}',
+        {%~ endfor %}
+    };
+
+    /**
+     * Set Endpoint
+     *
+     * Your project endpoint
+     *
+     * @param {string} endpoint
+     *
+     * @returns {this}
+     */
+    setEndpoint(endpoint: string): this {
+        this.config.endpoint = endpoint;
+        this.config.endpointRealtime = this.config.endpointRealtime || this.config.endpoint.replace('https://', 'wss://').replace('http://', 'ws://');
+
+        return this;
+    }
+
+    /**
+     * Set Realtime Endpoint
+     *
+     * @param {string} endpointRealtime
+     *
+     * @returns {this}
+     */
+    setEndpointRealtime(endpointRealtime: string): this {
+        this.config.endpointRealtime = endpointRealtime;
+
+        return this;
+    }
+
+    /**
+     * Set self-signed
+     *
+     * @param {boolean} selfSigned
+     *
+     * @returns {this}
+     */
+    setSelfSigned(selfSigned: boolean): this {
+        this.config.selfSigned = selfSigned;
+
+        return this;
+    }
+
+    /**
+     * Add header
+     *
+     * @param {string} header
+     * @param {string} value
+     *
+     * @returns {this}
+     */
+    addHeader(header: string, value: string): this {
+        this.headers[header.toLowerCase()] = value;
+
+        return this;
+    }
+
+    {%~ for header in spec.global.headers %}
+    /**
+     * Set {{header.key | caseUcfirst}}
+     *
+    {%~ if header.description %}
+     * {{header.description}}
+     *
+    {%~ endif %}
+     * @param value string
+     *
+     * @return {this}
+     */
+    set{{header.key | caseUcfirst}}(value: string): this {
+        this.headers['{{header.name}}'] = value;
+        this.config.{{ header.key | caseLower }} = value;
+        return this;
+    }
+    {%~ endfor %}
+
+    prepareRequest(method: string, url: URL, headers: Headers = {}, params: Payload = {}): { uri: string, options: RequestInit } {
+        method = method.toUpperCase();
+
+        headers = Object.assign({}, this.headers, headers);
+
+        let options: RequestInit = {
+            method,
+            headers,
+            credentials: 'include'
+        };
```

Comment:
I'm not sure we can achieve this without adding node dependency `https`

Reply:
Can we recommend just using NODE_TLS_REJECT_UNAUTHORIZED=0 instead?

Reply:
Or custom fetch

Reply:
How would they use a custom fetch?

Reply:
Added a setFetch method

Reply:
this going to make the node sdk work differently than any other SDK (because it's not as simple to allow self signed certificates in all runtime environments). Are we okay with this?

Reply:
Its much much better we keep everything consistent. @loks0n let's double check if we can handle this without a dependency. If we do add the dependency what implications does it have?

Reply:
@eldadfux, @lohanidamodar, no dependency is being added for this self-signed problem. The problem is the SDK needs to work in different runtimes like the edge or in react server components where either you can't allow a self-signed certificate or you would do it differently than if you were just in node.

Reply:
@stnguyen90 @loks0n can we keep the same behavior for node and just thrown and exception where our traditional setSelfSigned method is not supported?

Reply:
To implement `setSelfSigned` for node we'd have to `import { Agent } from 'undici'` or `import { Agent } from 'https'`

There is no way to conditionally import this (without dropping support for cjs & node 16), and therefore it would break on edge runtimes.

The alternative recommendation I'm suggesting is the dev would do this import for us:
```
import { Agent } from 'undici';

const client = new Client()
    .setProject(proces.env.MY_PROJECT_ID)
    .setFetch((uri, opts) => {
        const optsWithAgent = {
            ...opts,
            agent: new Agent({ rejectUnauthorised: false }) // disable ssl verification
        }
        return fetch(uri, optsWithAgent);
    });
    
 // ... make appwrite requests as normal

Reply:
Update: I was wrong there is a way. We now use forked version of `node-fetch-native` with this functionality